src/requirements/README.md

```diff
@@ -1,44 +1,103 @@
+as JSDoc directly in type definitions.
+
+The code for precomputing a list of fulfillable courses are in the [`checkers`](./checkers) folder.
+They are organized by colleges and majors. All of those checkers are aggregated together in the
+[`checkers/all-requirements.checker.ts`](./checkers/all-requirements-checkers.ts).
+
+## Architecture
+
+### Pre-computation phase
+
+To reiterate, this phase computes a list of satisfying courses for each requirement.
+
+1. Setup:
+   a. Each college or major requirement in
+   [`source-requirements-json.ts`](./source-requirements-json.ts) has a field called `checkerName`.
+   b. Each checker in
+   [`checkers/all-requirements.checker.ts`](./checkers/all-requirements-checkers.ts) is assigned
+   a checker name as its key.
+2. During pre-computation, we use the `checkerName` value in `source-requirements-json.ts` to find
+   a correct checker, and then run the checker. From that, we get a list of courses.
+3. We attach a list of courses to the requirement metadata in [`generator.ts`](./generator.ts).
+4. Now we created [`decorated-requirements.json`](./decorated-requirements.json).
+   Pre-computation is done.
+
+Each checker is either:
+
+- a TypeScript function `(course: Course) => boolean` that directly tells whether the course
+  satisfies the requirement that this checker correspond to.
+- or a list of such functions, where each function checks whether the course satisfies one of the
+  sub-requirements that this checker correspond to.
+
+This is how you should use your checkers:
+
+- Case 1: the requirement contains no sub-requirement. Use option 1. Example: CS major practicum.
+- Case 2: the requirement contains sub-requirements. Use option 2. Example: CS Core requirement.
+
+### Frontend Computation Phase
+
+#### University Requirements
+
+Since too many courses can satisfy these requirements, the satisfying course list is not
+pre-computed to reduce the data sent from server to client.
+
+It's computed on the client side directly in `computeUniversityRequirementFulfillments` function
+inside [`reqs-functions.ts`](./reqs-functions.ts).
+
+#### College or Major Requirements
+
+1. We first naively compute whether a sub-requirement is fulfilled by checking whether any course in
+   user's list is in the pre-computed satisfying course list.
+2. Then, we run a deduplication pass to remove all courses that have been double-counted that
+   should not be. (NOTE: THIS HAS NOT BEEN IMPLEMENTED YET.)
+3. Finally, We sum up the total number of sub-requirements/credits already fulfilled for each
+   requirement, to decide whether a requirement as a whole as been satisfied.
+
+## Contributing
+
+If you want to change a requirement checker, edit relevant files in `checkers` folder.
+
+If you want to add a support for new major or college, add relevant files in `checkers` folder.
+Then, import and link your new checkers in
```

Comment:
Could you add a little bit more information here, specifically on what "import and link your new checkers" entails? Since we might have some people less experienced with requirements adding data, it would be useful to expand on this.

Reply:
Added examples